src/pygeos.c

```diff
@@ -0,0 +1,252 @@
+  for (i = 0; i < n1; i++) {
+    ring1 = GEOSGetInteriorRingN_r(ctx, geom1, i);
+    if (ring1 == NULL) {
+      return 2;
+    }
+    ring2 = GEOSGetInteriorRingN_r(ctx, geom2, i);
+    if (ring2 == NULL) {
+      return 2;
+    }
+    ret = equals_identical_simple(ctx, ring1, ring2);
+    if (ret != 1) {
+      return ret;
+    }
+  }
+  return 1;
+}
+
+/* Note: forward declaration of the final function that is defined further below,
+so that it can be used in equals_identical_collection just below (due to the
+recursive nature of the implementation). */
+char equals_identical(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                      const GEOSGeometry* geom2);
+
+/* Check the sub-geometries of Multi-part geometries.
+Returns 1 on true, 0 on false, 2 on exception */
+char equals_identical_collection(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                                 const GEOSGeometry* geom2) {
+  int i, n1, n2;
+  const GEOSGeometry *sub_geom1 = NULL;
+  const GEOSGeometry *sub_geom2 = NULL;
+  char ret;
+
+  n1 = GEOSGetNumGeometries_r(ctx, geom1);
+  if (n1 == -1) {
+    return 2;
+  }
+  n2 = GEOSGetNumGeometries_r(ctx, geom2);
+  if (n2 == -1) {
+    return 2;
+  }
+  if (n1 != n2) {
+    return 0;
+  }
+
+  /* validate the sub geometries */
+  for (i = 0; i < n1; i++) {
+    sub_geom1 = GEOSGetGeometryN_r(ctx, geom1, i);
+    if (sub_geom1 == NULL) {
+      return 2;
+    }
+    sub_geom2 = GEOSGetGeometryN_r(ctx, geom2, i);
+    if (sub_geom2 == NULL) {
+      return 2;
+    }
+    ret = equals_identical(ctx, sub_geom1, sub_geom2);
+    if (ret != 1) {
+      return ret;
+    }
+  }
+  return 1;
+}
+
+/* Check geometries for being identical.
+Returns 1 on true, 0 on false, 2 on exception */
+char equals_identical(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                      const GEOSGeometry* geom2) {
+  int type1, type2;
+
+  type1 = GEOSGeomTypeId_r(ctx, geom1);
+  if (type1 == -1) {
+    return 2;
+  }
+  type2 = GEOSGeomTypeId_r(ctx, geom2);
+  if (type2 == -1) {
+    return 2;
+  }
+  if (type1 != type2) {
+    return 0;
+  }
+
+  if ((type1 == 0) || (type1 == 1) || (type1 == 2)) {
+    return equals_identical_simple(ctx, geom1, geom2);
+  } else if (type1 == 3) {
+    return equals_identical_polygon(ctx, geom1, geom2);
+  } else if ((type1 >= 4) && (type1 <= 7)) {
+    return equals_identical_collection(ctx, geom1, geom2);
+  } else {
+    return 2;
+  }
+}
+
+/*
+ * Determine pointwise equivalence of two geometries by checking
+ * that the structure, ordering, and values of all vertices are
+ * identical in all dimensions. NaN values are considered to be
+ * equal to other NaN values in the same position within coordinate sequences.
+ * \returns 1 on true, 0 on false, 2 on exception
+ */
+char PyGEOSEqualsIdentical(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                           const GEOSGeometry* geom2) {
+  return equals_identical(ctx, geom1, geom2);
```

Comment:
What is the purpose of having both `PyGEOSEqualsIdentical` and `equals_identical` with the same signature when `PyGEOSEqualsIdentical` just calls `equals_identical` ?
Maybe use alias instead if you wanted two names for the same function? https://stackoverflow.com/questions/40968732/how-to-give-alias-for-a-function-in-c